src/metatensor/models/soap_bpnn/train.py

```diff
@@ -123,27 +128,83 @@ def train(
+        aggregated_validation_info: Dict[str, Tuple[float, int]] = {}
+
         train_loss = 0.0
         for batch in train_dataloader:
             optimizer.zero_grad()
             structures, targets = batch
-            loss = compute_model_loss(loss_fn, model, structures, targets)
+            loss, info = compute_model_loss(loss_fn, model, structures, targets)
             train_loss += loss.item()
             loss.backward()
             optimizer.step()
+            aggregated_train_info = update_aggregated_info(aggregated_train_info, info)
+        aggregated_train_info = finalize_aggregated_info(aggregated_train_info)
 
         validation_loss = 0.0
         for batch in validation_dataloader:
             structures, targets = batch
             # TODO: specify that the model is not training here to save some autograd
-            loss = compute_model_loss(loss_fn, model, structures, targets)
+            loss, info = compute_model_loss(loss_fn, model, structures, targets)
             validation_loss += loss.item()
+            aggregated_validation_info = update_aggregated_info(
+                aggregated_validation_info, info
+            )
+        aggregated_validation_info = finalize_aggregated_info(
+            aggregated_validation_info
+        )
 
+        # Now we log the information:
         if epoch % hypers_training["log_interval"] == 0:
-            logger.info(
-                f"Epoch {epoch}, train loss: {train_loss:.4f}, "
-                f"validation loss: {validation_loss:.4f}"
+            logging_string = (
+                f"Epoch {epoch:4}, train loss: {train_loss:10.4f}, "
+                " validation loss: {validation_loss:10.4f}"
             )
+            for key, value in aggregated_train_info.items():
+                if key.endswith("_positions_gradients"):
+                    # check if this is a force
+                    target_name = key[: -len("_positions_gradients")]
+                    if model.capabilities.outputs[target_name].quantity == "energy":
+                        # if this is a force, replace the ugly name with "force"
+                        if only_one_energy:
+                            key = "force"
+                        else:
+                            key = f"force[{target_name}]"
+                elif key.endswith("_displacement_gradients"):
+                    # check if this is a virial/stress
+                    target_name = key[: -len("_displacement_gradients")]
+                    if model.capabilities.outputs[target_name].quantity == "energy":
+                        # if this is a virial/stress,
+                        # replace the ugly name with "virial/stress"
+                        if only_one_energy:
+                            key = "virial/stress"
+                        else:
+                            key = f"virial/stress[{target_name}]"
+                logging_string += f", train {key} RMSE: {value:10.4f}"
```

Comment:
This block and the one below seem to me exactly the same. Maybe just do another outer loop?

Reply:
Very good idea